internal/v5/auth_test.go

```diff
@@ -295,6 +302,45 @@ var readAuthorizationTests = []struct {
 		Code:    params.ErrUnauthorized,
 		Message: `unauthorized: access denied for user "kirk"`,
 	},
+}, {
+	about:               "access provided through development channel",
+	username:            "kirk",
+	groups:              []string{"group1", "group2", "group3"},
+	readPerm:            []string{"picard", "sisko", "group42", "group47"},
+	developmentReadPerm: []string{"group1"},
+	channels:            []charmstore.Channel{charmstore.DevelopmentChannel},
+}, {
+	about:               "access provided through development channel, but charm not published",
+	username:            "kirk",
+	groups:              []string{"group1", "group2", "group3"},
+	readPerm:            []string{"picard", "sisko", "group42", "group47"},
+	developmentReadPerm: []string{"group1"},
+	expectStatus:        http.StatusUnauthorized,
+	expectBody: params.Error{
+		Code:    params.ErrUnauthorized,
+		Message: `unauthorized: access denied for user "kirk"`,
+	},
+}, {
+	about:               "access provided through stable channel",
+	username:            "kirk",
+	groups:              []string{"group1", "group2", "group3"},
+	readPerm:            []string{"picard", "sisko", "group42", "group47"},
+	developmentReadPerm: []string{"group12"},
+	stableReadPerm:      []string{"group2"},
+	channels:            []charmstore.Channel{charmstore.DevelopmentChannel, charmstore.StableChannel},
+}, {
+	about:               "access provided through stable channel, but charm not published",
+	username:            "kirk",
+	groups:              []string{"group1", "group2", "group3"},
+	readPerm:            []string{"picard", "sisko", "group42", "group47"},
+	developmentReadPerm: []string{"group12"},
+	stableReadPerm:      []string{"group2"},
+	channels:            []charmstore.Channel{charmstore.DevelopmentChannel},
+	expectStatus:        http.StatusUnauthorized,
+	expectBody: params.Error{
+		Code:    params.ErrUnauthorized,
+		Message: `unauthorized: access denied for user "kirk"`,
+	},
```

Comment:
If we keep propagating unpublished ACLs to devel and stable, we should test those cases as well.
If not, we should test that having unpublished access does not grant development or stable authorization.